// vite-plugin-ssr/dist/cjs/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__private = exports._injectAssets = exports.pipeNodeStream = exports.pipeWebStream = exports.dangerouslySkipEscape = exports.escapeInject = exports.createPageRenderer = void 0;
require("./page-files/setup");
var createPageRenderer_1 = require("./createPageRenderer");
Object.defineProperty(exports, "createPageRenderer", { enumerable: true, get: function () { return createPageRenderer_1.createPageRenderer; } });
var escapeInject_1 = require("./html/escapeInject");
Object.defineProperty(exports, "escapeInject", { enumerable: true, get: function () { return escapeInject_1.escapeInject; } });
Object.defineProperty(exports, "dangerouslySkipEscape", { enumerable: true, get: function () { return escapeInject_1.dangerouslySkipEscape; } });
var stream_1 = require("./html/stream");
Object.defineProperty(exports, "pipeWebStream", { enumerable: true, get: function () { return stream_1.pipeWebStream; } });
Object.defineProperty(exports, "pipeNodeStream", { enumerable: true, get: function () { return stream_1.pipeNodeStream; } });
var injectAssets_1 = require("./html/injectAssets");
Object.defineProperty(exports, "_injectAssets", { enumerable: true, get: function () { return injectAssets_1.injectAssets; } });
const importBuild_1 = require("./importBuild");
exports.__private = { importBuild: importBuild_1.importBuild };
// Plugin should be imported from `vite-plugin-ssr/plugin`
const utils_1 = require("../shared/utils");
exports.default = pluginWrongImport;
function pluginWrongImport() {
    (0, utils_1.assertUsage)(false, "`import ssr from 'vite-plugin-ssr';` is depecrated: use `import ssr from 'vite-plugin-ssr/plugin';` instead.");
}
// Enable `const ssr = require('vite-plugin-ssr')`
// This lives at the end of the file to ensure it happens after all assignments to `exports`
module.exports = Object.assign(exports.default, exports);
